### PR TITLE
Accessibility: Add Landmark navigation

### PR DESCRIPTION
**Pull request by akirk**
This adds macro keyboard navigation between the sections of Element Web, so you can cycle through these items:
- Space Panel: active space
- Room Search
- Room List: active room
- Message Composer

Fixes https://github.com/element-hq/element-web/issues/22674

landmark-navigation

Screenshot 2023-12-08 at 08 06 40

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Signed-off-by: `Alex Kirk <akirk@users.noreply.github.com>`

---
This PR currently has none of the required changelog labels.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.

### CODE REVIEW

File: src/components/structures/LeftPanel.tsx [hunk 1]
```diff
@@ -309,6 +309,22 @@ export default class LeftPanel extends React.Component<IProps, IState> {
...
         }
+        const navAction = getKeyBindingsManager().getNavigationAction(ev);
+        switch (navAction) {
+            case KeyBindingAction.NextLandmark:
+                ev.stopPropagation();
+                ev.preventDefault();
+                (
+                    document.querySelector(".mx_RoomTile_selected") ||
+                    document.querySelector(".mx_RoomListHeader_contextMenuButton")
+                )?.focus();
```
Comment (t3chguy) [inline src/components/structures/LeftPanel.tsx:320]:
```suggestion
                (
                    document.querySelector<HTMLElement>(".mx_RoomTile_selected") ??
                    document.querySelector<HTMLElement>(".mx_RoomListHeader_contextMenuButton")
                )?.focus();
```

Or maybe would the following work

```suggestion
                document.querySelector<HTMLElement>(".mx_RoomTile_selected, .mx_RoomListHeader_contextMenuButton"?.focus();
```

A comment would go a long way too so it can be maintained appropriately, why these two classes

File: src/components/structures/LeftPanel.tsx [hunk 2]
```diff
@@ -309,6 +309,22 @@ export default class LeftPanel extends React.Component<IProps, IState> {
...
+                break;
+            case KeyBindingAction.PreviousLandmark:
+                ev.stopPropagation();
+                ev.preventDefault();
+                document.querySelector(".mx_SpaceButton_active")?.focus();
```
Comment (t3chguy) [inline src/components/structures/LeftPanel.tsx:325]:
```suggestion
                document.querySelector<HTMLElement>(".mx_SpaceButton_active")?.focus();
```

Comment would be appreciated

File: src/accessibility/KeyboardShortcuts.ts [hunk 3]
```diff
@@ -120,6 +120,10 @@ export enum KeyBindingAction {
     SelectPrevUnreadRoom = "KeyBinding.previousUnreadRoom",
     /** Select next room with unread messages */
     SelectNextUnreadRoom = "KeyBinding.nextUnreadRoom",
+    /** Select prev message */
+    SelectPrevMessage = "KeyBinding.previousMessage",
+    /** Select next message */
+    SelectNextMessage = "KeyBinding.nextMessage",
```
Comment (t3chguy) [inline src/accessibility/KeyboardShortcuts.ts:126]: These seem unrelated to landmark navigation

File: src/accessibility/KeyboardShortcuts.ts [hunk 4]
```diff
@@ -286,10 +290,14 @@ export const CATEGORIES: Record<CategoryName, ICategory> = {
             KeyBindingAction.SelectPrevUnreadRoom,
             KeyBindingAction.SelectNextRoom,
             KeyBindingAction.SelectPrevRoom,
+            KeyBindingAction.SelectNextMessage,
+            KeyBindingAction.SelectPrevMessage,
```
Comment (t3chguy) [inline src/accessibility/KeyboardShortcuts.ts:294]: This is also stray code from the other PR

File: src/accessibility/KeyboardShortcuts.ts [hunk 5]
```diff
@@ -558,6 +566,20 @@ export const KEYBOARD_SHORTCUTS: IKeyboardShortcuts = {
...
+        },
+        displayName: _td("keyboard|next_message"),
+    },
+    [KeyBindingAction.SelectPrevMessage]: {
+        default: {
+            ctrlOrCmdKey: true,
+            key: Key.ARROW_UP,
+        },
+        displayName: _td("keyboard|prev_message"),
+    },
```
Comment (t3chguy) [inline src/accessibility/KeyboardShortcuts.ts:582]: This is also stray code from the other PR

File: src/components/structures/MessagePanel.tsx [hunk 6]
```diff
@@ -422,6 +425,42 @@ export default class MessagePanel extends React.Component<IProps, IState> {
...
+                defaultDispatcher.dispatch(
+                    {
+                        action: Action.FocusSendMessageComposer,
+                        context: TimelineRenderingType.Room,
+                    },
+                    true,
+                );
+            }
+        }
+
```
Comment (t3chguy) [inline src/components/structures/MessagePanel.tsx:463]: This is also stray code from the other PR

File: src/components/views/rooms/EventTile.tsx [hunk 7]
```diff
@@ -1211,6 +1211,7 @@ export class UnwrappedEventTile extends React.Component<EventTileProps, IState>
                         "ref": this.ref,
                         "className": classes,
                         "aria-live": ariaLive,
+                        "aria-label": this.props.mxEvent.sender.name + ". " + this.props.mxEvent.event.content.body,
```
Comment (t3chguy) [inline src/components/views/rooms/EventTile.tsx:1214]: This is also stray code from the other PR

File: src/components/views/rooms/EventTile.tsx [hunk 8]
```diff
@@ -1396,6 +1397,7 @@ export class UnwrappedEventTile extends React.Component<EventTileProps, IState>
                         "className": classes,
                         "tabIndex": -1,
                         "aria-live": ariaLive,
+                        "aria-label": this.props.mxEvent.sender.name + ". " + this.props.mxEvent.event.content.body,
```
Comment (t3chguy) [inline src/components/views/rooms/EventTile.tsx:1400]: This is also stray code from the other PR

File: src/components/views/rooms/SendMessageComposer.tsx [hunk 9]
```diff
@@ -308,6 +308,17 @@ export class SendMessageComposer extends React.Component<ISendMessageComposerPro
...
+        switch (navAction) {
+            case KeyBindingAction.SelectPrevMessage:
+                (
+                    document.querySelector(".mx_EventTile_selected") || document.querySelector(".mx_EventTile_last")
+                )?.focus();
+                event.preventDefault();
+                event.stopPropagation();
+                return;
+        }
+
```
Comment (t3chguy) [inline src/components/views/rooms/SendMessageComposer.tsx:321]: This is also stray code from the other PR

File: src/components/structures/LeftPanel.tsx [hunk 10]
```diff
@@ -309,6 +309,24 @@ export default class LeftPanel extends React.Component<IProps, IState> {
...
+        const navAction = getKeyBindingsManager().getNavigationAction(ev);
+        switch (navAction) {
+            case KeyBindingAction.NextLandmark:
+                ev.stopPropagation();
+                ev.preventDefault();
+                // The next landmark is the selected room or the room first room, if none selected.
+                (
+                    document.querySelector<HTMLElement>(".mx_RoomTile_selected") ||
+                    document.querySelector<HTMLElement>(".mx_RoomTile")
+                )?.focus();
```
Comment (t3chguy) [inline src/components/structures/LeftPanel.tsx:321]: Any chance we could have a single class/something which is responsible for the overarching order of landmarks in this navigation so you don't have to read a bunch of files to maintain it. Like an enum for the various landmarks and a function with signature `navigateLandmark(currentLandmark: EnumType, backward: boolean): void` or similar

**Review by t3chguy — COMMENTED** [review]
Looks generally sane, other than tests

File: test/accessibility/KeyboardLandmarkUtils-test.tsx [hunk 11]
```diff
@@ -0,0 +1,131 @@
+/*
```
[the remark this answers is not in the document]
Comment (MidhunSureshR) [reply test/accessibility/KeyboardLandmarkUtils-test.tsx:1]: I copied the header from another test so this also happens elsewhere:

image

I'll fix those in another PR.

**Comment by MidhunSureshR** [comment]
I think we should ignore sonarcloud because any new code with low coverage is in `onKeyDown` and that path is sufficiently tested by the playwright test.

File: src/accessibility/LandmarkNavigation.ts [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+ * The landmarks are cycled through in the following order:
+ * ACTIVE_SPACE_BUTTON <-> ROOM_SEARCH <-> ROOM_LIST <-> MESSAGE_COMPOSER/HOME <-> ACTIVE_SPACE_BUTTON
+ */
+export class LandmarkNavigation {
+    private static getIndexInArray(i: number): number {
+        const n = ORDERED_LANDMARKS.length;
+        // calculating index in the circular array
+        const nextIndex = ((i % n) + n) % n;
+        return nextIndex;
+    }
```
Comment (t3chguy) [inline src/accessibility/LandmarkNavigation.ts:49]: Could probably simplify it using https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at and a modulo

File: src/accessibility/LandmarkNavigation.ts [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+                },
+                true,
+            );
+            return true;
+        } else {
+            const e = document.querySelector<HTMLElement>(".mx_HomePage");
+            e?.focus();
+            return !!e;
+        }
+    },
```
Comment (t3chguy) [inline src/accessibility/LandmarkNavigation.ts:137]: Couldn't the signature be that it returns the `e` and the caller focuses it if `!undefined` otherwise moves onto the next one? would get rid of this 2 line boilerplate in every focusFunction

File: src/components/views/rooms/BasicMessageComposer.tsx [hunk 14]
```diff
@@ -536,6 +537,17 @@ export default class BasicMessageEditor extends React.Component<IProps, IState>
...
+        const navAction = getKeyBindingsManager().getNavigationAction(event);
+        switch (navAction) {
+            case KeyBindingAction.PreviousLandmark:
+                LandmarkNavigation.navigateToPreviousLandmarkFrom(Landmark.MESSAGE_COMPOSER_OR_HOME);
+                handled = true;
+                break;
+            case KeyBindingAction.NextLandmark:
+                LandmarkNavigation.navigateToNextLandmarkFrom(Landmark.MESSAGE_COMPOSER_OR_HOME);
+                handled = true;
+                break;
```
Comment (t3chguy) [inline src/components/views/rooms/BasicMessageComposer.tsx:549]: a `findAndFocusNextLandmark(enumValue, navAction === KeyBinding.PreviousLandmark)` would simplify this boilerplate in every receiver class

**Comment by t3chguy** [comment]
> I think we should ignore sonarcloud because any new code with low coverage is in onKeyDown and that path is sufficiently tested by the playwright test.

Fine, the playwright test looks comprehensive

File: src/accessibility/LandmarkNavigation.ts (hunk 12 above)
Comment (MidhunSureshR) [reply src/accessibility/LandmarkNavigation.ts]: https://github.com/matrix-org/matrix-react-sdk/pull/12190/commits/b2ce73b04ce066b4d325afcee4ca36d6ab56c74f

File: src/accessibility/LandmarkNavigation.ts (hunk 13 above)
Comment (MidhunSureshR) [reply src/accessibility/LandmarkNavigation.ts]: https://github.com/matrix-org/matrix-react-sdk/pull/12190/commits/ea4fea8b96af0c010ce4598b04459d7bc3046549

File: src/components/views/rooms/BasicMessageComposer.tsx (hunk 14 above)
Comment (MidhunSureshR) [reply src/components/views/rooms/BasicMessageComposer.tsx]: https://github.com/matrix-org/matrix-react-sdk/pull/12190/commits/ea4fea8b96af0c010ce4598b04459d7bc3046549

File: src/accessibility/LandmarkNavigation.ts [hunk 15]
```diff
@@ -68,55 +60,31 @@ export class LandmarkNavigation {
...
-        return !!e;
-    },
-    [Landmark.ROOM_LIST]: () => {
-        const e =
-            document.querySelector<HTMLElement>(".mx_RoomTile_selected") ||
-            document.querySelector<HTMLElement>(".mx_RoomTile");
-        e?.focus();
-        return !!e;
-    },
+const landmarkToDOMElementMap = {
```
Comment (t3chguy) [inline src/accessibility/LandmarkNavigation.ts:80]:
```suggestion
const landmarkToDomElementMap = {
```

Sorry being pedantic: https://www.approxion.com/capital-offenses-how-to-handle-abbreviations-in-camelcase/

File: src/components/structures/LeftPanel.tsx [hunk 16]
```diff
@@ -309,18 +309,15 @@ export default class LeftPanel extends React.Component<IProps, IState> {
...
-                ev.preventDefault();
-                LandmarkNavigation.navigateToPreviousLandmarkFrom(Landmark.ROOM_SEARCH);
-                break;
+        if (navAction) {
+            ev.stopPropagation();
+            ev.preventDefault();
+            LandmarkNavigation.findAndFocusNextLandmark(
+                Landmark.ROOM_SEARCH,
+                navAction === KeyBindingAction.PreviousLandmark,
+            );
```
Comment (t3chguy) [inline src/components/structures/LeftPanel.tsx:320]: I think this should check it is one of the expected 2 keybinding actions, otherwise other navigation bindings would also cause landmarks to move: https://github.com/matrix-org/matrix-react-sdk/blob/c843a4163a4f03dbca2bc78ffb1ed64c247e53b9/src/accessibility/KeyboardShortcuts.ts#L277-L295

Comment (MidhunSureshR) [reply src/components/structures/LeftPanel.tsx]: https://github.com/matrix-org/matrix-react-sdk/pull/12190/commits/114cc77acdcc8b9527e139710e162a951896d28b

File: src/accessibility/LandmarkNavigation.ts (hunk 15 above)
Comment (MidhunSureshR) [reply src/accessibility/LandmarkNavigation.ts]: https://github.com/matrix-org/matrix-react-sdk/pull/12190/commits/0961caea83d25b7d70febb49871e63f0ec6b10de and https://github.com/matrix-org/matrix-react-sdk/pull/12190/commits/1d8ab2a1bc886d44e8f0ddc0d955da7281a43193

**Review by t3chguy — COMMENTED** [review]
Code looks great now, testing it on Netlify looks like the focus outline isn't showing up though which makes it really difficult to use


https://github.com/user-attachments/assets/5c1f6dc1-a3b4-4111-8c7b-e09b4e7be935

**Comment by MidhunSureshR** [comment]
hmm.. it shows up in firefox but not in chromium

**Comment by MidhunSureshR** [comment]
@t3chguy 

This is a chrome issue; from https://blog.chromium.org/2020/09/giving-users-and-developers-more.html:

>Because mouse users may frequently use keyboard shortcuts, Chrome's implementation will bypass "keyboard mode" if a meta key (such as command, control, etc.) is pressed. For example, if a user who was previously using a mouse pressed a keyboard shortcut which shows a settings dialog, :focus-visible would not match on the focused element in the settings dialog.

[focusVisible](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus#focusvisible) option should fix this in the future.

I'm partial to merging this as is but also happy to write a `focusWithIndicator` helper if needed.

**Comment by t3chguy** [comment]
That makes a lot of sense, could you add focusVisible anyway, as a best effort

File: src/accessibility/LandmarkNavigation.ts [hunk 17]
```diff
@@ -18,6 +18,12 @@ import { TimelineRenderingType } from "../contexts/RoomContext";
 import { Action } from "../dispatcher/actions";
 import defaultDispatcher from "../dispatcher/dispatcher";
 
+/**
+ * In future, browsers will support focusVisible option.
+ * See https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus#focusvisible
+ */
+type ExperimentalFocusOptions = FocusOptions & { focusVisible: true };
```
Comment (t3chguy) [inline src/accessibility/LandmarkNavigation.ts:25]: Can you add this type in `global.d.ts` instead to avoid needing to type cast and to get the benefit in all files?

Comment (MidhunSureshR) [reply src/accessibility/LandmarkNavigation.ts]: done, https://github.com/matrix-org/matrix-react-sdk/pull/12190/commits/d1085b7b8ef877d1481788b849c22ce805461b33

**Review by t3chguy — APPROVED** [review]
:shipit:

File: src/accessibility/LandmarkNavigation.ts (hunk 17 above)
Comment (t3chguy) [reply src/accessibility/LandmarkNavigation.ts]: It might need to be optional `focusVisible?: boolean` but CI will show
